### PR TITLE
Set vertex data fixes

### PR DESCRIPTION
**Pull request by PatrickvL**
This builds upon and extends RadWolfie's solution, fixing #1385.

Test case : BeginPush XDK sample.

(One path is untested, thus marked with LOG_TEST_CASE)

### CODE REVIEW

File: src/CxbxKrnl/EmuD3D8.cpp [hunk 1]
```diff
@@ -4093,14 +4096,48 @@ VOID WINAPI XTL::EMUPATCH(D3DDevice_SetVertexData4f)
...
+			X_D3DVertexShader *pXboxVertexShader = VshHandleToXboxVertexShader(g_CurrentXboxVertexShaderHandle);
+			if (!(pXboxVertexShader->Flags & 0x10/*=X_VERTEXSHADER_PROGRAM*/)) {
+				ActiveVertexAttributeFlags = pXboxVertexShader->Flags;
+			}
+		}
+
+		// Handle persistent vertex attribute flags, by resetting non-persistent colors
+		// to their default value (and leaving the persistent colors alone - see the
+		// "Copy all attributes of the previous vertex" comment below) :
+		static const uint32_t ColorBlack = D3DCOLOR_ARGB(0, 0, 0, 0);
```
Comment (RadWolfie) [inline src/CxbxKrnl/EmuD3D8.cpp]: Incorrect, it is suppose to be 255 (Alpha), 0 (Red), 0 (Green), 0 (Blue). That's the correct define as black.

P.S. If setting Alpha to 0, then any color is not visible.

File: src/CxbxKrnl/EmuD3D8.cpp [hunk 2]
```diff
@@ -4093,10 +4096,45 @@ VOID WINAPI XTL::EMUPATCH(D3DDevice_SetVertexData4f)
...
-		// TODO : Handle D3DUSAGE_PERSISTENTDIFFUSE, D3DUSAGE_PERSISTENTSPECULAR,
-		// D3DUSAGE_PERSISTENTBACKDIFFUSE, and D3DUSAGE_PERSISTENTBACKSPECULAR 
+
+		// Get the vertex shader flags (if any is active) :
+		uint32_t ActiveVertexAttributeFlags = 0;
+		if (VshHandleIsVertexShader(g_CurrentXboxVertexShaderHandle)) {
+			LOG_TEST_CASE("D3DDevice_SetVertexData4f with active VertexShader");
+			X_D3DVertexShader *pXboxVertexShader = VshHandleToXboxVertexShader(g_CurrentXboxVertexShaderHandle);
+			if (!(pXboxVertexShader->Flags & 0x10/*=X_VERTEXSHADER_PROGRAM*/)) {
+				ActiveVertexAttributeFlags = pXboxVertexShader->Flags;
```
Comment (RadWolfie) [inline src/CxbxKrnl/EmuD3D8.cpp]: 👍 I don't think I would ever find out how to use the current shader into a "X_D3DVertexShader" pointer. :wink:

File: src/CxbxKrnl/EmuD3D8.cpp (hunk 1 above)
Comment (PatrickvL) [reply src/CxbxKrnl/EmuD3D8.cpp]: True, except D3D seems to set zero too

**Comment by RadWolfie** [comment]
Since it is getting accurate. I'm merging this pull request.

**Pull request merged**